### PR TITLE
机器之魂：聊天机器人是怎么工作的

### LINKED ISSUE #2008: 机器之魂：聊天机器人是怎么工作的
* 原文链接：[Soul of the Machine: How Chatbots Work](https://medium.com/@gk_/how-chat-bots-work-dfff656a35e2)
  * Markdown文件：[文件地址](https://github.com/xitu/gold-miner/blob/master/TODO/how-chat-bots-work.md)
  * PR 地址：#2011
  * 文章分类：**AI**
  ----
  * 翻译时间：5 天
  * 校对时间：1 天
  * 翻译积分：5 分
  * 校对积分：1 分
  * [积分有什么用](https://github.com/xitu/gold-miner/wiki)

### PR DESCRIPTION
**Pull request by lsvih**
close #2008

### CODE REVIEW

**Comment by lsvih** [comment]
#2008

**Comment by lsvih** [comment]
看来必须要在 pr 的说明里就 reference issue 才行 XD

**Comment by lsvih** [comment]
看来只能创建 pr 的时候就 reference issue，创建完再修改也识别不了了 XD

**Comment by lileizhenshuai** [comment]
@sqrthree  校对认领

**Comment by sqrthree** [comment]
@lileizhenshuai 好的呢 :beer:

File: TODO/how-chat-bots-work.md [hunk 1]
```diff
@@ -1,166 +1,175 @@
...
 
-    Human: Do you know who Albert Einstein is
-    Robot: Albert Einstein was a German physicist.
+````
+Human: Do you know who Albert Einstein is
+Robot: Albert Einstein was a German physicist.
+````
 
-It knows who a physicist is only because his or her name has an associated pattern. Likewise it responds to anything solely because of an authored pattern. Given hundreds or thousands of patterns you might see a chatbot “persona” emerge.
+它之所以知道别人问的是哪个物理学家，只是靠着与他或者她名字相关联的模式匹配。同样的，它靠着创作者模式可以对任何意图进行回应。在给予它成千上万种模式之后，你终将能看到一个“类人”的聊天机器人出现。
```
Comment (lileizhenshuai) [inline TODO/how-chat-bots-work.md]: 它靠着创作者模式可以对任何意图进行回应 ->
它靠着创作者预设的模式可以对任何意图进行回应

File: TODO/how-chat-bots-work.md [hunk 2]
```diff
@@ -1,166 +1,175 @@
...
+````
 
-    input: "What’s it like outside?"
-     term: "it" **(class: weather (2), greeting)**
-     term: "outside **(class: weather (2) )**
-     classification: **weather **(score=4)
+请注意，“What’s it like outside”在分类时找到了另一个分类的单词，但是正确的分类给了单词较高的分值。通过算法公式，我们可以为句子计算匹配每个分类对应的词频，因此不需要去标明所有的模式。
 
-Notice that the classification for “What’s it like outside” found a term in another class but the term similarities to the desired class produced a higher score. By using an equation we are looking for word matches given some sample sentences for each class, and we avoid having to identify every pattern.
+这种分类器通过标定分类分值（计算词频）的方法给出最匹配语句的分类，但是它仍然有局限性。分值与概率不同，它仅仅能告诉我们句子的意图最有可能是哪个分类，而不能告诉我们它的所有匹配分类的可能性。因此，很难去给出一个阈值来判定是接受这个得分结果还是不接受这个结果。这种类型的算法给出的最高分仅仅能作为判断相关性的基础，本质上作为分类器还是比较差的。此外，这个算法不能接受 *is not* 类型的句子，因为它仅仅计算了 *it* 可能是啥。也就是说这种方法不适合做为包含 *not* 的否定句的分类。
```
Comment (lileizhenshuai) [inline TODO/how-chat-bots-work.md]: 本质上作为分类器还是比较差的 ->
它本质上作为分类器的效果还是比较差的

Comment (lileizhenshuai) [inline TODO/how-chat-bots-work.md]: 因为它仅仅计算了 *it* 可能是啥 ->
因为它仅仅计算了 *it* 可能是什么

File: TODO/how-chat-bots-work.md [hunk 3]
```diff
@@ -1,166 +1,175 @@
...
 
-As in the prior method, each class is given with some number of example sentences. Once again each sentence is broken down by word (stemmed) and each word becomes an input for the neural network. The synaptic weights are then calculated by iterating through the training data thousands of times, each time adjusting the weights slightly to greater accuracy. By recalculating back across multiple layers (“back-propagation”) the weights of all synapses are calibrated while the results are compared to the training data output. These weights are like a ‘strength’ measure, in a neuron the synaptic weight is what causes something to be more memorable than not. You remember a thing more because you’ve seen it more times: each time the ‘weight’ increases slightly.
+在前面的方法里，每个分类都会给定一些例句。接着，根据词干进行分句，将所有单词作为神经网络的输入。然后遍历数据，进行成千上万次迭代计算，每次迭代都通过改变突触权重来得到更高的准确率。接着通过反过来通过对训练集输出值和神经网络计算结果的对比，对各层重新进行计算权重（反向传播）。这个“权重”可以类比成神经突触想记住某个东西的“力度”，你能记住某个东西是因为你曾多次见过它，在每次见到它的时候这个“权重”都会轻微地上升。
 
-At some point the adjustment reaches a point of diminishing returns, this is called “over-fitting” and going beyond this is counter-productive.
+有时，在权重调整到某个程度后反而会使得结果逐渐变差，这种情况称为“过拟合”，在出现过拟合的情况下继续进行训练会导致反效果。
```
Comment (lileizhenshuai) [inline TODO/how-chat-bots-work.md]: 在出现过拟合的情况下继续进行训练会导致反效果 ->
在出现过拟合的情况下继续进行训练，反而会适得其反

File: TODO/how-chat-bots-work.md [hunk 4]
```diff
@@ -1,166 +1,175 @@
...
 
 ![](https://cdn-images-1.medium.com/max/1600/1*QckgibgJ74BhMaqinqwSDw.png)
 
-The trained neural network is less code than an comparable algorithm but it requires a potentially large matrix of “weights”. In a relatively small sample, where the training sentences have 150 unique words and 30 classes this would be a matrix of 150x30. Imagine multiplying a matrix of this size 100,000 times to establish a sufficiently low error rate. This is where processing speed comes in.
+训练好的神经网络模型的代码量其实很小，不过它需要一个很大的潜在权重矩阵。举个相对较小的样例，它的训练句子包括了 150 个单词、30 种分类，这可能产生一个 150x30 大小的矩阵；你可以想象一下，为了降低错误率，这么大的一个矩阵需要反复的进行 10 万次矩阵乘法。这也是为啥说需要高性能处理器的原因。
```
Comment (lileizhenshuai) [inline TODO/how-chat-bots-work.md]: 这也是为啥说 ->
这也是为什么说
更加书面化一点

File: TODO/how-chat-bots-work.md [hunk 5]
```diff
@@ -1,166 +1,175 @@
...
 
-If the neural network sounds magnificently sophisticated, relax, it boils down to [matrix multiplication](https://www.khanacademy.org/math/precalculus/precalc-matrices/multiplying-matrices-by-matrices/v/matrix-multiplication-intro) and a [formula for reducing values](https://en.wikipedia.org/wiki/Sigmoid_function) between -1 and 1 or some other minimal range. A middle-school math student could learn this in a few hours. The hard work is achieving clean training data.
+神经网络之所以能够做到既复杂又稀疏，归结于[矩阵乘法](https://www.khanacademy.org/math/precalculus/precalc-matrices/multiplying-matrices-by-matrices/v/matrix-multiplication-intro)和一种[缩小值至 -1，1 区间的公式](https://en.wikipedia.org/wiki/Sigmoid_function)（即激活函数，这里指的是 Sigmoid），一个中学生也能在几小时内学会它。其实真正困难的工作是清洗训练数据。_
```
Comment (lileizhenshuai) [inline TODO/how-chat-bots-work.md]: 一个中学生也能在几小时内学会它。其实真正困难的工作是清洗训练数据。_ ->
一个中学生也能在几小时内学会它。其实真正困难的工作是清洗训练数据。

File: TODO/how-chat-bots-work.md [hunk 6]
```diff
@@ -1,166 +1,175 @@
...
-Just as there are variations in pattern matching code and in algorithms, there are variations in neural networks, some more complex than others. The basic machinery is the same. The essential work is that of classification.
+就像前面的模式匹配和算法匹配一样，神经网络也有各种各样的变体，有一些变体会十分复杂。不过它的基本原理是相同的，做的主要工作也都是进行分类。
 
 ![](https://cdn-images-1.medium.com/max/1600/1*_ldEr2WurmqNq6Pgp5J24w.jpeg)
 
-The mechanical music box knows nothing about music theory, likewise **chatbot machinery knows nothing about language**.
+机械音乐盒并不了解乐理，同样的，**聊天机器人并不了解语言**。
 
-Chatbot machinery is looking for patterns in collections of terms, each term is reduced to a token. In this machine* words have no meaning* except for their patterned existence within training data. The label “artificial intelligence” applied to such machinery [is mostly BS](https://medium.com/@gk_/the-ai-label-is-bullshit-559b171867ff#.3tlhftemt).
+聊天机器人实质上就是寻找短语集合中摸的模式，每个短语还能再分割成单个单词。在聊天机器人内部，除了它们存在的模式以及训练数据之外的**单词其实并没有意义**。为这样的“机器人”贴上“人工智能”的标签其实[也很糟糕](https://medium.com/@gk_/the-ai-label-is-bullshit-559b171867ff#.3tlhftemt)。
```
Comment (lileizhenshuai) [inline TODO/how-chat-bots-work.md]: 聊天机器人实质上就是寻找短语集合中摸的模式 ->
聊天机器人实质上就是寻找短语集合中的模式

File: TODO/how-chat-bots-work.md [hunk 7]
```diff
@@ -1,166 +1,175 @@
...
-An early example of automation — the mechanical music box.
-A set of tuned metallic teeth aligned on a comb-like structure are positioned next to a cylinder with pins. Each pin corresponds to a note at a specific interval.
+早期自动化的样例 —— 机械音乐盒。
+一组经过调音的金属齿排列成梳状结构，置于一个有针的圆柱边上。每根针都以一个特定的时间对应着一个音符。
 
-As the mechanism turns, it creates a tune by striking one or multiple pins at the predefined time. To play a different song, you can swap in a different cylindrical drum (assuming the set of unique notes is equivalent).
+当机械转动时，它便会在预定好的时间通过单个或者多个针的拨动来产生乐曲。如果要播放不同的歌，你得换不同的圆柱桶（假设不同的乐曲对应的特定音符是一样的）。
 
-In addition to striking a note, the movement of the drum can cause other automation, such as a moving figurine. Either way, the fundamental machinery of the music box remains the same.
+除了发出音符之外，圆桶的转动还可以造成一些其它的动作，例如移动小雕像等。不管怎样，这个音乐盒的基本机械结构是不会变的。
```
Comment (lileizhenshuai) [inline TODO/how-chat-bots-work.md]: 我去查了一下“圆筒”和“圆桶”的区别，似乎“桶”侧重用来盛放东西，有底；“筒”则是类似竹子一样的结构。
所以这里可能用圆筒（和最后一段）一样比较好？

File: TODO/how-chat-bots-work.md [hunk 8]
```diff
@@ -1,166 +1,175 @@
...
   ![](https://cdn-images-1.medium.com/max/2000/1*HRgcOpW8vSPqM-GxkoHhWw.jpeg)
 
-Since the early industrial age, we’ve been fascinated by self-operating devices. They represent the humanization of technology.
+自早期的工业时代以来，人类就被能自主操作的设备迷住了。因为，它们代表了科技的“人化”。
 
-Today, it is software that that’s becoming more human — most obviously “chatbots.”
+而在今天，各种软件也在逐渐变得人性化。其中变化最明显的当属“聊天机器人”。
 
-But how do these machines work? First, wind back time and explore an earlier — yet similar — technology.
+但是这些“机械”是如何运作的呢？首先，让我们回溯过去，探寻一种原始，但典型的技术。
```
Comment (lileizhenshuai) [inline TODO/how-chat-bots-work.md]: 首先，让我们回溯过去，探寻一种原始，但典型的技术。 ->
首先，让我们回溯过去，探寻一种原始，但相似的技术。
觉得整篇文章的核心是音乐盒和聊天机器人原理类似，所以这边可能用相似更好？

**Comment by lileizhenshuai** [comment]
@sqrthree  一校完成

**Comment by jasonxia23** [comment]
@sqrthree 认领校对～

**Comment by sqrthree** [comment]
@jasonxia23 妥妥哒 :beers:

**Comment by lsvih** [comment]
@jasonxia23  hi，大佬有空的时候来校对一下哈

**Comment by jasonxia23** [comment]
@lsvih 昨晚已经校对一半了，今天团建，马上好，稍等哈

**Comment by lsvih** [comment]
@jasonxia23 没事没事= =就提醒您一下。。现在改了也得等根号三上班才发-。-

File: TODO/how-chat-bots-work.md [hunk 9]
```diff
@@ -1,166 +1,175 @@
...
+有许多的聊天机器人框架[都是用这种方法来判断意图分类](https://medium.com/@gk_/text-classification-using-algorithms-e4d50dcba45#.ewnhttxa4)。而且大多数都是针对训练集进行词频计算，这种“幼稚”的方法有时还意外的有效。
 
-Many chatbot frameworks use [algorithms such as this to classify intent](https://medium.com/@gk_/text-classification-using-algorithms-e4d50dcba45#.ewnhttxa4). Most of what’s taking place is word counting against training datasets, it’s “naive” but surprisingly effective.
 
-### Neural Networks
+### 神经网络
 
-Artificial neural networks, invented in the 1940’s, are a way of calculating an output from an input (a classification) using weighted connections (“synapses”) that are calculated from repeated iterations through training data. Each pass through the training data alters the weights such that the neural network produces the output with greater “accuracy” (lower error rate).
+人工神经网络发明于 19 世纪 40 年代，它通过迭代计算训练数据得到连接的加权值（“突触”）,然后用于对输入数据进行分类。通过一次次使用训练数据计算改变加权值以使得神经网络的输出得到更高的“准确率”（低错误率）。
```
Comment (jasonxia23) [inline TODO/how-chat-bots-work.md]: 19 世纪
=>
20 世纪

Comment (jasonxia23) [inline TODO/how-chat-bots-work.md]: （“突触”）,
=>
（“突触”），

File: TODO/how-chat-bots-work.md [hunk 10]
```diff
@@ -1,166 +1,175 @@
...
+上图为一种神经网络结构，其中包括神经元（圆）和突触（线）
+
+其实除了当今的软件可以用更快的处理器、更大的内存外，这些结构并没有出现什么新奇的东西。当做数十万次的矩阵乘法（神经网络中的基本数学运算）的时候，运行内存和计算速度成为了关键问题。
 
-As in the prior method, each class is given with some number of example sentences. Once again each sentence is broken down by word (stemmed) and each word becomes an input for the neural network. The synaptic weights are then calculated by iterating through the training data thousands of times, each time adjusting the weights slightly to greater accuracy. By recalculating back across multiple layers (“back-propagation”) the weights of all synapses are calibrated while the results are compared to the training data output. These weights are like a ‘strength’ measure, in a neuron the synaptic weight is what causes something to be more memorable than not. You remember a thing more because you’ve seen it more times: each time the ‘weight’ increases slightly.
+在前面的方法里，每个分类都会给定一些例句。接着，根据词干进行分句，将所有单词作为神经网络的输入。然后遍历数据，进行成千上万次迭代计算，每次迭代都通过改变突触权重来得到更高的准确率。接着通过反过来通过对训练集输出值和神经网络计算结果的对比，对各层重新进行计算权重（反向传播）。这个“权重”可以类比成神经突触想记住某个东西的“力度”，你能记住某个东西是因为你曾多次见过它，在每次见到它的时候这个“权重”都会轻微地上升。
```
Comment (jasonxia23) [inline TODO/how-chat-bots-work.md]: 接着通过反过来通过
通过重复哈

**Comment by lsvih** [comment]
@sqrthree 根据校对意见修改完毕

**Comment by sqrthree** [comment]
@lsvih 已经 merge 啦~ 快快麻溜发布到掘金专栏然后给我发下链接，方便及时添加积分哟。
